cmd/algokey/keyreg.go

```diff
@@ -0,0 +1,254 @@
+			network,
+			strings.Join(validNetworkList, ", "))
+	}
+
+	return gen, nil
+}
+
+func run(params keyregCmdParams) error {
+	// Implicit last valid
+	if params.lastValid == 0 {
+		params.lastValid = params.firstValid + txnLife
+	}
+
+	if !params.offline {
+		if params.partkeyFile == "" {
+			return errors.New("must provide --keyfile when registering participation keys")
+		}
+		if params.addr != "" {
+			return errors.New("do not provide --address when registering participation keys")
+		}
+	} else {
+		if params.addr == "" {
+			return errors.New("must provide --address when bringing an account offline")
+		}
+		if params.partkeyFile != "" {
+			return errors.New("do not provide --keyfile when bringing an account offline")
+		}
+	}
+
+	var accountAddress basics.Address
+	if params.addr != "" {
+		var err error
+		accountAddress, err = basics.UnmarshalChecksumAddress(params.addr)
+		if err != nil {
+			return fmt.Errorf("unable to parse --address: %w", err)
+		}
+	}
+
+	if params.partkeyFile != "" && !util.FileExists(params.partkeyFile) {
+		return fmt.Errorf("cannot access keyfile '%s'", params.partkeyFile)
+	}
+
+	if params.txFile == "" {
+		params.txFile = fmt.Sprintf("%s.tx", params.partkeyFile)
+	}
+
+	if util.FileExists(params.txFile) || params.txFile == stdoutFilenameValue {
+		return fmt.Errorf("outputFile '%s' already exists", params.partkeyFile)
+	}
+
+	// Lookup information from partkey file
+	var part *account.Participation
+	if params.partkeyFile != "" {
+		partDB, err := db.MakeErasableAccessor(params.partkeyFile)
+		if err != nil {
+			return fmt.Errorf("cannot open keyfile %s: %v", params.partkeyFile, err)
+		}
+
+		partkey, err := account.RestoreParticipation(partDB)
+		if err != nil {
+			return fmt.Errorf("cannot load keyfile %s: %v", params.partkeyFile, err)
+		}
+		defer partkey.Close()
+
+		part = &partkey.Participation
+
+		if params.firstValid < uint64(part.FirstValid) {
+			return fmt.Errorf("firstvalid (%d) should be greater than or equal to the first valid participation key (%d)", params.firstValid, part.FirstValid)
+		}
+	}
+
+	validRange := params.lastValid - params.firstValid
+	if validRange > txnLife {
+		return fmt.Errorf("firstvalid (%d) is %d greater than lastvalid (%d). lastvalid may not be greater than firstvalid + %d", params.firstValid, validRange, params.lastValid, txnLife)
+	}
+
+	var txn transactions.Transaction
+	if !params.offline {
+		// Generate go-online transaction
+		txn = part.GenerateRegistrationTransaction(
+			basics.MicroAlgos{Raw: params.fee},
+			basics.Round(params.firstValid),
+			basics.Round(params.lastValid),
+			[32]byte{},
+			part.StateProofSecrets != nil)
+	} else {
+		// Generate go-offline transaction
+		txn = transactions.Transaction{
+			Type: protocol.KeyRegistrationTx,
+			Header: transactions.Header{
+				Sender:     accountAddress,
+				Fee:        basics.MicroAlgos{Raw: params.fee},
```

Comment:
what will happen if a user inputs fee=0? 
according to the code on `MakeUnsignedGoOnlineTx` (https://github.com/algorand/go-algorand/blob/6ad1e686172bf967b72fb11f6657c1f5b9f78fb6/libgoal/transactions.go#L243),
when fee is 0 then we calculate it as non-flat-fee (which we can't do cause we don't have an online node)
my question is: should we reject fee=0?

Reply:
Yes, it's always treated as a flat fee. There is a missing check to make sure the `fee < minFee`, I'll add that.